server/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java

```diff
@@ -290,30 +305,42 @@ protected AnalyzedRelation visitJoin(Join node, StatementAnalysisContext stateme
                         relationContext.parentSources(),
                         coordinatorTxnCtx.sessionSettings().searchPath().currentSchema()),
                     new SubqueryAnalyzer(this, statementContext));
-                Expression expr;
-                if (joinCriteria instanceof JoinOn) {
-                    expr = ((JoinOn) joinCriteria).getExpression();
-                } else {
-                    expr = JoinUsing.toExpression(
-                        leftRel.relationName().toQualifiedName(),
-                        rightRel.relationName().toQualifiedName(),
-                        ((JoinUsing) joinCriteria).getColumns());
-                }
-                try {
-                    joinCondition = expressionAnalyzer.convert(expr, relationContext.expressionAnalysisContext());
-                } catch (RelationUnknown e) {
-                    throw new RelationValidationException(e.getTableIdents(),
-                        String.format(Locale.ENGLISH,
-                        "missing FROM-clause entry for relation '%s'", e.getTableIdents()));
-                }
-            } else {
-                throw new UnsupportedOperationException(String.format(Locale.ENGLISH, "join criteria %s not supported",
-                    joinCriteria.getClass().getSimpleName()));
+                var expressionAnalysisContext = statementContext.currentRelationContext().expressionAnalysisContext();
+                joinCondition = expressionAnalyzer.convert(expression, expressionAnalysisContext);
+            } catch (RelationUnknown e) {
+                throw new RelationValidationException(
+                    e.getTableIdents(),
+                    String.format(Locale.ENGLISH, "missing FROM-clause entry for relation '%s'", e.getTableIdents())
+                );
             }
         }
+        var joinType = JoinType.values()[node.getType().ordinal()];
+        JoinRelation joinRelation = new JoinRelation(leftRel, rightRel, leftRel.outputs(), joinType, joinCondition);
+        JoinPair joinPair = extractJoinPair(joinRelation, statementContext.currentRelationContext().sourceNames());
+        statementContext.currentRelationContext().addJoinPair(joinPair);
+        return joinRelation;
+    }
 
-        relationContext.addJoinType(JoinType.values()[node.getType().ordinal()], joinCondition);
-        return null;
+    private static JoinPair extractJoinPair(JoinRelation joinRelation, List<RelationName> relationsInQueryOrder) {
+        assert relationsInQueryOrder.size() >= 2 : "sources must be added first, cannot add join pair for only 1 source";
+
+        var relations = new ArrayList<>(relationsInQueryOrder);
```

Comment:
This copy is needed, because `relationsInQueryOrder` is immutable, and may need to be modified.